Adding quickTunerStat.py

ethansaurusrex opened the pull request:
Adding statistics and testing script `quickTunerStat.py`, accepts `*.qtfiles` from `quickTunerGen.py`. Can either compare against data frame created by `quickTunerStat.py` or by running `rocmlir-tuning-driver` (currently disabled).

ethansaurusrex committed:
Changing input format

ethansaurusrex committed:
Check

ethansaurusrex committed:
rewrite for use of rocmlir-tuning-driver

ethansaurusrex committed:
Working data verifier

ethansaurusrex committed:
Fixing tuning runner portions

ethansaurusrex committed:
Checking in, rewrote some of the base class functions

ethansaurusrex committed:
Adding statistics and testing script quickTunerStat.py, accepts *.qt
files from quickTunerGen.py. Can either compare against dataframe
created by quickTunerStat.py or by running rocmlir-tuning-driver
(currently disabled).

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,632 @@
+
+Generated statistics and verify quick tuning configs
+
+optional arguments:
```
Why's there a copy of `usage` here? Shouldn't `argparse` generate it? Some examples might be good

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,632 @@
+
+        if match:
+            tup = match.groups()
+            transA = True if tup[0].lower() == 'true' else False
```
The `GemmConfig` class over in `tuningRunner` already has this and it might be less fragile?

ethansaurusrex replied:
The issue is I still need to create a tuple in order to index the dictionary contained within the class. I switched out the method of converting the transA/B to the same method used in `tuningRunner`/`perfRunner`

krzysz00 replied:
I'm more suggesting using the classes in tuningRunner or perfRunner to parse the perf config - I think `parameterSweeps.py` might be a good place to look or refactor

ethansaurusrex replied:
Of course I will take a look at that

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,632 @@
+            tile_params = tile_params.drop(['param8','param9'], axis=1)
+
+            tile_params['performance'] = df['NormalizedTFlops']
+
```
Nit: a bunch of extra newlines that don't have semantic meaning, please remove?

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,632 @@
+        self.output_df = pd.DataFrame(rank_dict)
+        print(self.output_df)
+
+class tunerValidator(perfConfigValidator):
```
Nit: upper case class names

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,632 @@
+                 rocmlir_path,
+                 rocm_build_script='/share/scripts/build-rocm'):
+        """
+        initializer
```
Not worth a doc string

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,632 @@
+        
+        self.cpp_file = os.path.join(os.path.dirname(rocmlir_path), self.gridwise_gemm_params)
+        self.rocm_build_script = rocm_build_script
+        self.backup = self.cpp_file + ".bu"
```
`~` is the usual suffix for backup files, no?

ethansaurusrex committed:
Updated script with debug information and switched indexing
to 'NormalizedTFlops' for extracting performance information
when validating with --data flag. Cleaned up --tuning code,
still disabled until further testing is done.

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,672 @@
+
+
+    def __del__(self):
+        # copy back original file
```
Hm?

krzysz00 reviewed (commented):
I could do with a better picture of what's going on here and how I'm meant to operate this

ethansaurusrex committed:
Fixed capitalized class names

ethansaurusrex committed:
Removed unneeded newlines, comments, and methods

ethansaurusrex committed:
Fixed backup filename

ethansaurusrex committed:
Changed method for converting transA/B string to boolean. Also
updated column names, correcting previous splitK mistake.

ethansaurusrex committed:
Adding usage README.md and fixed some issues with column naming.
Added cpp file output on invocation of --rank

ethansaurusrex committed:
Fixed dtype lookup bug

ethansaurusrex committed:
Added tflops back in to df

ethansaurusrex committed:
Replaced manual parsing of config string with
perfRunner.GemmConfiguration.fromCommandLine()

ethansaurusrex committed:
Updated README.md to include data collection method using tuna-script.sh

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,565 @@
+from dataclasses import dataclass
+from perfCommonUtils import Operation
+import perfRunner
+from sklearn.preprocessing import MinMaxScaler
```
We might want to add `sklearn` to a Dockerfile or a requirements.txt somewhere?

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,565 @@
+        # to validate file we need data already read,
+
+        all_data = []
+        columns = ['M/block', 'N/block', 'K/block', 'M/wave', 'N/wave', 'kPack', 'splitK', 'forceUnroll', 'bCopyMore']
```
I'll note that, for example, the attention perf configs don't have this structure, and, more interestingl, that f32 gemm/conv on Navi doesn't either. But this is probably fine, just wanted to call out the limitation

krzysz00 commented on mlir/utils/performance/quickTunerStat.py:
```diff
@@ -0,0 +1,565 @@
+        with open(cpp_filename, 'w') as file:
+            file.write(cpp_content)
+
+    def buildRocm(self):
```
buildRocmlir, maybe?

Also, why is this here?

krzysz00 reviewed (approved):
I'm rather confused as to when I'd want the "rebuild rocMLIR" mode in a script that's meant to produce statistics.

That being said, ... approved on the assumption that you'll explain this somewhere at some point

ethansaurusrex committed:
Changed romclir building during rocmlir tuner validation

ethansaurusrex committed:
Quick checkin

ethansaurusrex committed:
Fixed tuner method build method

ethansaurusrex committed:
Fixed single file and directory input ranking

ethansaurusrex committed:
Changed format for input to match quickTunerGen.py

ethansaurusrex committed:
Added convolution

djramic commented:
Closed as we have new script: #1689
